Fixed collection sorting error

afbora opened the pull request:
## Describe the PR
This fix is used for cases when main sorting specified by columns has equal values
Without it it will lead to `Fatal Error: Nesting level too deep - recursive dependency?`

Tested with Kirby 3.2.3

Waiting your reviews.

## Related issues

- Fixes https://github.com/getkirby/kirby/issues/1790

## Reference solution
https://github.com/yiisoft/yii2/commit/78de3157942dcbae43c0f5172bc009b17025e319

## Todos

- [x] Add unit tests for fixed bug/feature
- [x] Pass all unit tests
- [x] Fix code style issues with CS fixer and `composer fix`
- [x] If needeed, in-code documentation (DocBlocks etc.)

afbora committed:
Fixed collection sorting error

Fix recursive dependency error on sorting collection

#2020

afbora commented on src/Toolkit/Collection.php:
```diff
@@ -936,12 +936,33 @@ public function sortBy()
-            $params[] = $field['values']    ?? [];
+            $values = [];
+            if (empty($field['values']) === false) {
+                $values = array_filter($field['values']) ? array_values($field['values']) : array_keys($field['values']);
```
This is for stable work. For ex: in files collection array values are empty, but pages collection array keys and values are filled.

afbora commented on src/Toolkit/Collection.php:
```diff
@@ -936,12 +936,33 @@ public function sortBy()
+                $values = array_filter($field['values']) ? array_values($field['values']) : array_keys($field['values']);
+                
+                // check array has duplicate values
+                if (count($values) !== count(array_unique($values))) {
```
This issue about same values and we should check array has duplicated values.

afbora commented on src/Toolkit/Collection.php:
```diff
@@ -936,12 +936,33 @@ public function sortBy()
 
+        // This fix is used for cases when main sorting specified by columns has equal values
+        // Without it it will lead to Fatal Error: Nesting level too deep - recursive dependency?
+        if ($hasDuplicateValues === true) {
```
Now we will only use this new feature if array/data has duplicated values

Btw if we don't check the duplicates, the sorting will be wrong.

lukasbestle reviewed (commented):
Thanks a lot for tracking down this bug and finding a potential solution (also thanks for linking to the source, that's very helpful!).

Unfortunately I can't reproduce it myself, so I can't really test if this fixes it.

One general question though: Is it really necessary to first check if the data has duplicate values? Because that makes the whole logic a *lot* more complex and unpredictable. Also I imagine that this change will be quite a performance hit for large data sets as every value list for every attribute/field needs to be checked for duplicates.
My thinking is: If we add the additional "artificial" sorting attribute as the last attribute, it should only ever be used by `array_multisort` if all prior attributes are duplicates. So it shouldn't matter if it's there if it's not needed.

You write that the sorting will be wrong if the check is omitted – wrong in what way and why?

lukasbestle commented on src/Toolkit/Collection.php:
```diff
@@ -908,7 +908,13 @@ public function sortBy()
                     // make sure that we return something sortable
                     // but don't convert other scalars (especially numbers) to strings!
-                    $values[$key] = is_scalar($value) === true ? $value : (string)$value;
+                    $fieldValue = is_scalar($value) === true ? $value : (string)$value;
```
In your commit description you write:

> This is for stable work. For ex: in files collection array values are empty, but pages collection array keys and values are filled.

To be honest I don't really understand why this change is needed. Is this related to the "Nesting level too deep - recursive dependency?" error or to a different issue? If so, please send me some details on that issue and why this change solves it.

afbora replied:
https://github.com/getkirby/kirby/pull/2020#issuecomment-531459943

lukasbestle commented on src/Toolkit/Collection.php:
```diff
@@ -922,7 +928,13 @@ public function sortBy()
                     // make sure that we return something sortable
                     // but don't convert other scalars (especially numbers) to strings!
-                    $values[$key] = is_scalar($value) === true ? $value : (string)$value;
+                    $fieldValue = is_scalar($value) === true ? $value : (string)$value;
```
Just for my own overview when reviewing this again later: Same change as above.

lukasbestle reviewed (commented):
Thank you very much! 👍❤️

I have modified the code to work for a few more edge cases according to my proposal in https://github.com/getkirby/kirby/pull/2020#issuecomment-524652356.

I have also changed the unit tests to use only mock objects. There are two reasons for that:

1. We want to decouple the Toolkit (and other non-CMS) tests from the CMS as our other code packages may be used outside of a CMS installations. The tests should therefore be able to run without the `Cms` classes being present. It also improves the reliability of the tests as less code is tested at once.
2. Your tests were creating the actual pages in the file system and therefore couldn't run more than once without manual intervention (deleting the pages again). We generally use a `fixtures` directory inside the test directories for that. That directory is then cleaned up automatically after each test. But it's even better to write the tests in a way that they don't require changes to the file system at all (like with the mock classes).

Sorry that I had to throw away the tests you wrote, but you couldn't know the first reason, so it's no problem at all. :)

---

Now only my question about your "Rewrite loop for collection data values" commit and after that we should be ready to go. 🎉

afbora commented on src/Toolkit/Collection.php:
```diff
@@ -908,7 +908,13 @@ public function sortBy()
-                    $values[$key] = is_scalar($value) === true ? $value : (string)$value;
+                    $fieldValue = is_scalar($value) === true ? $value : (string)$value;
+
+                    if (strlen($fieldValue) > 0) {
```
As you know this mean:

    if ($fieldValue != "") {

lukasbestle committed:
Explicit sorting by filename to fix wrong order

If the `sort` field is not set for some or all files, the sorting would be arbitrary. By sorting by filename as a fallback, there is always a predictable order.

#2020

lukasbestle reviewed (approved):
This is now ready for 3.2.5. Thanks again @afbora, your help was amazing, we wouldn't have found the issue and a solution without you. 👍

@bastianallgeier @distantnative We need to keep an eye on this during the RC phase, maybe explicitly ask the users to test sorting in their sites. There are quite a few edge-cases here and we might have missed some with the fix.

bastianallgeier committed:
Merge branch 'develop' into fix/1790-collection-recursive-dependency-error
